File: src/formatters.ts
// Convert Atlassian Document Format to readable plain text
export function adfToText(adf: any): string {
  if (!adf || !adf.content) return "";
  return adf.content.map(nodeToText).join("\n\n");
}

function nodeToText(node: any): string {
  switch (node.type) {
    case "paragraph":
      return inlineToText(node.content);
    case "heading": {
      const level = node.attrs?.level || 1;
      return "#".repeat(level) + " " + inlineToText(node.content);
    }
    case "bulletList":
      return (node.content || [])
        .map((item: any) => "- " + listItemText(item))
        .join("\n");
    case "orderedList":
      return (node.content || [])
        .map((item: any, i: number) => `${i + 1}. ` + listItemText(item))
        .join("\n");
    case "codeBlock": {
      const lang = node.attrs?.language || "";
      return "```" + lang + "\n" + inlineToText(node.content) + "\n```";
    }
    case "blockquote":
      return (node.content || [])
        .map((c: any) => "> " + nodeToText(c))
        .join("\n");
    case "rule":
      return "---";
    case "table":
      return tableToText(node);
    default:
      return inlineToText(node.content);
  }
}

function inlineToText(content: any[]): string {
  if (!content) return "";
  return content
    .map((c: any) => {
      if (c.type === "text") return c.text || "";
      if (c.type === "hardBreak") return "\n";
      if (c.type === "mention") return `@${c.attrs?.text || c.attrs?.id || "unknown"}`;
      if (c.type === "emoji") return c.attrs?.shortName || "";
      if (c.type === "inlineCard") return c.attrs?.url || "";
      return c.text || "";
    })
    .join("");
}

function listItemText(item: any): string {
  return (item.content || []).map(nodeToText).join("\n  ");
}

function tableToText(node: any): string {
  const rows = (node.content || []).map((row: any) =>
    (row.content || [])
      .map((cell: any) => (cell.content || []).map(nodeToText).join(" "))
      .join(" | ")
  );
  return rows.join("\n");
}

// Convert plain text to ADF for Jira API v3
export function textToAdf(text: string) {
  return {
    type: "doc",
    version: 1,
    content: text.split("\n").map((line) => ({
      type: "paragraph" as const,
      content: line ? [{ type: "text" as const, text: line }] : [],
    })),
  };
}

// Format an issue object into readable text
export function formatIssue(issue: any): string {
  const f = issue.fields || {};
  const lines: string[] = [
    `${issue.key}: ${f.summary || "(no summary)"}`,
    `  Status: ${f.status?.name || "Unknown"}`,
    `  Type: ${f.issuetype?.name || "Unknown"}`,
  ];

  if (f.priority) lines.push(`  Priority: ${f.priority.name}`);
  if (f.assignee) lines.push(`  Assignee: ${f.assignee.displayName} (${f.assignee.accountId})`);
  if (f.reporter) lines.push(`  Reporter: ${f.reporter.displayName}`);
  if (f.labels?.length) lines.push(`  Labels: ${f.labels.join(", ")}`);
  if (f.components?.length)
    lines.push(`  Components: ${f.components.map((c: any) => c.name).join(", ")}`);
  if (f.fixVersions?.length)
    lines.push(`  Fix Versions: ${f.fixVersions.map((v: any) => v.name).join(", ")}`);
  if (f.sprint) lines.push(`  Sprint: ${f.sprint.name}`);
  if (f.parent) lines.push(`  Parent: ${f.parent.key}`);
  if (f.created) lines.push(`  Created: ${f.created}`);
  if (f.updated) lines.push(`  Updated: ${f.updated}`);
  if (f.resolution) lines.push(`  Resolution: ${f.resolution.name}`);

  if (f.description) {
    const desc = typeof f.description === "string" ? f.description : adfToText(f.description);
    lines.push(`  Description:\n    ${desc.split("\n").join("\n    ")}`);
  }

  return lines.join("\n");
}

// Format an issue as a compact one-liner for search results
export function formatIssueSummary(issue: any): string {
  const f = issue.fields || {};
  const assignee = f.assignee?.displayName || "Unassigned";
  return `${issue.key}  [${f.status?.name}]  ${f.summary}  (${f.issuetype?.name}, ${assignee})`;
}

File: jira_mcp/formatters.py
from __future__ import annotations


# --- ADF to plain text ---

def adf_to_text(adf: dict | None) -> str:
    if not adf or not adf.get("content"):
        return ""
    return "\n\n".join(_node_to_text(node) for node in adf["content"])


def _node_to_text(node: dict) -> str:
    t = node.get("type")
    if t == "paragraph":
        return _inline_to_text(node.get("content"))
    if t == "heading":
        level = (node.get("attrs") or {}).get("level", 1)
        return "#" * level + " " + _inline_to_text(node.get("content"))
    if t == "bulletList":
        return "\n".join("- " + _list_item_text(item) for item in (node.get("content") or []))
    if t == "orderedList":
        return "\n".join(
            f"{i + 1}. " + _list_item_text(item)
            for i, item in enumerate(node.get("content") or [])
        )
    if t == "codeBlock":
        lang = (node.get("attrs") or {}).get("language", "")
        return f"```{lang}\n{_inline_to_text(node.get('content'))}\n```"
    if t == "blockquote":
        return "\n".join("> " + _node_to_text(c) for c in (node.get("content") or []))
    if t == "rule":
        return "---"
    if t == "table":
        return _table_to_text(node)
    return _inline_to_text(node.get("content"))


def _inline_to_text(content: list | None) -> str:
    if not content:
        return ""
    parts = []
    for c in content:
        t = c.get("type")
        if t == "text":
            parts.append(c.get("text", ""))
        elif t == "hardBreak":
            parts.append("\n")
        elif t == "mention":
            attrs = c.get("attrs") or {}
            parts.append(f"@{attrs.get('text') or attrs.get('id', 'unknown')}")
        elif t == "emoji":
            parts.append((c.get("attrs") or {}).get("shortName", ""))
        elif t == "inlineCard":
            parts.append((c.get("attrs") or {}).get("url", ""))
        else:
            parts.append(c.get("text", ""))
    return "".join(parts)


def _list_item_text(item: dict) -> str:
    return "\n  ".join(_node_to_text(c) for c in (item.get("content") or []))


def _table_to_text(node: dict) -> str:
    rows = []
    for row in node.get("content") or []:
        cells = [
            " ".join(_node_to_text(c) for c in (cell.get("content") or []))
            for cell in (row.get("content") or [])
        ]
        rows.append(" | ".join(cells))
    return "\n".join(rows)


# --- Plain text to ADF ---

def text_to_adf(text: str) -> dict:
    return {
        "type": "doc",
        "version": 1,
        "content": [
            {
                "type": "paragraph",
                "content": [{"type": "text", "text": line}] if line else [],
            }
            for line in text.split("\n")
        ],
    }


# --- Structured issue extraction ---

def extract_issue(issue: dict) -> dict:
    """Extract a structured dict from a raw Jira issue response."""
    f = issue.get("fields") or {}
    result = {
        "key": issue["key"],
        "summary": f.get("summary"),
        "status": (f.get("status") or {}).get("name"),
        "type": (f.get("issuetype") or {}).get("name"),
    }
    if f.get("priority"):
        result["priority"] = f["priority"]["name"]
    if f.get("assignee"):
        a = f["assignee"]
        result["assignee"] = {"displayName": a["displayName"], "accountId": a["accountId"]}
    if f.get("reporter"):
        result["reporter"] = f["reporter"]["displayName"]
    if f.get("labels"):
        result["labels"] = f["labels"]
    if f.get("components"):
        result["components"] = [c["name"] for c in f["components"]]
    if f.get("fixVersions"):
        result["fixVersions"] = [v["name"] for v in f["fixVersions"]]
    if f.get("sprint"):
        result["sprint"] = f["sprint"]["name"]
    if f.get("parent"):
        result["parent"] = f["parent"]["key"]
    if f.get("duedate"):
        result["duedate"] = f["duedate"]
    if f.get("startDate"):
        result["startDate"] = f["startDate"]
    if f.get("created"):
        result["created"] = f["created"]
    if f.get("updated"):
        result["updated"] = f["updated"]
    if f.get("resolution"):
        result["resolution"] = f["resolution"]["name"]
    if f.get("description"):
        result["description"] = (
            f["description"] if isinstance(f["description"], str)
            else adf_to_text(f["description"])
        )
    return result


def extract_issue_summary(issue: dict) -> dict:
    """Extract a compact structured dict from a Jira issue."""
    f = issue.get("fields") or {}
    return {
        "key": issue["key"],
        "summary": f.get("summary"),
        "status": (f.get("status") or {}).get("name"),
        "type": (f.get("issuetype") or {}).get("name"),
        "assignee": (f.get("assignee") or {}).get("displayName"),
        "priority": (f.get("priority") or {}).get("name"),
    }
